### PR TITLE
feat: CE-1045 add action taken filter for ceeb

### PR DESCRIPTION
**Pull request by mikevespi**
# Description

Added a filter on Action Taken to the complaints list filters for CEEB users. If the filter is is present the search function will send a request to case management for a list of `complaint_identifier`s (in CM this is found through case_file -> lead_identifier via the new `getLeadsByActionTaken`). These are then used to filter the complaints via an AND WHERE...IN condition. If no `lead_identifier`s are returned, then an array of `[-1]` is passed in its place, since WHERE...IN() causes an error in postgres with an empty set.

Fixes # (issue)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

Testing included manually testing the filter in both the complaint list, and map view. A Cypress test was added, however at the time of writing this there is no seed data in Case Management, so setup for the test (assigning an action taken to a complaint so that there is something to show up once the filter is applied) takes place in the test itself, which is not ideal.

When manually testing, if no results show after filtering, clear the filter, go into one of the actions and set an action taken, then try again. 

## Checklist

- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Further comments

---

Thanks for the PR!

Deployments, as required, will be available below:
- [Frontend](https://nr-compliance-enforcement-707-frontend.apps.silver.devops.gov.bc.ca)
- [Backend](https://nr-compliance-enforcement-707-frontend.apps.silver.devops.gov.bc.ca/api)

Please create PRs in draft mode.  Mark as ready to enable:
- [Analysis Workflow](https://github.com/bcgov/nr-compliance-enforcement/actions/workflows/analysis.yml)

After merge, new images are deployed in:
- [Merge Workflow](https://github.com/bcgov/nr-compliance-enforcement/actions/workflows/merge.yml)

---

Thanks for the PR!

Deployments, as required, will be available below:
- [Frontend](https://nr-compliance-enforcement-708-frontend.apps.silver.devops.gov.bc.ca)
- [Backend](https://nr-compliance-enforcement-708-frontend.apps.silver.devops.gov.bc.ca/api)

Please create PRs in draft mode.  Mark as ready to enable:
- [Analysis Workflow](https://github.com/bcgov/nr-compliance-enforcement/actions/workflows/analysis.yml)

After merge, new images are deployed in:
- [Merge Workflow](https://github.com/bcgov/nr-compliance-enforcement/actions/workflows/merge.yml)

### CODE REVIEW

File: frontend/src/app/components/containers/complaints/complaint-filter.tsx [hunk 1]
```diff
@@ -22,6 +23,8 @@ import { ComplaintFilterPayload, updateFilter } from "../../../store/reducers/co
 import Option from "../../../types/app/option";
 import { getUserAgency } from "../../../service/user-service";
 import { listActiveFilters } from "../../../store/reducers/app";
+import UserService from "../../../service/user-service";
```
Comment (afwilcox) [inline frontend/src/app/components/containers/complaints/complaint-filter.tsx:26]: As per the sonar cloud comment, the previous import of { getUserAgency } is no longer required.

File: backend/src/v1/complaint/complaint.service.ts [hunk 2]
```diff
@@ -871,6 +873,30 @@ export class ComplaintService {
...
+          throw new Error("GraphQL errors occurred");
+        }
+        /**
+         * If no leads in the case manangement database have had the selected action taken, `getLeadsByActionTaken`
+         * returns an empty array, and WHERE...IN () does not accept an empty set, it throws an error. In our use
+         * case, if `getLeadsByActionTaken` returns an empty array, we do not want the entire search to error, it
+         * should simply return an empty result set. To handle this, if `getLeadsByActionTaken` returns an empty
+         * array, we populate the array with a value that would never match on a complaint_identifier, -1.
+         */
+        const complaintIdentifiers = data.getLeadsByActionTaken.length > 0 ? data.getLeadsByActionTaken : [-1];
```
Comment (afwilcox) [inline backend/src/v1/complaint/complaint.service.ts:893]: thank you for this detailed comment!

File: frontend/cypress/e2e/complaint-search.v2.cy.ts [hunk 3]
```diff
@@ -146,3 +147,45 @@ describe("Complaint Search Functionality", () => {
...
+  it("allows filtering of complaints by Action Taken", function () {
+    // Navigate to the complaint list
+    const complaintWithActionTakenID = "23-030990";
+    const actionTaken = "Forward to lead agency";
+    cy.visit("/");
+    cy.waitForSpinner();
+
+    // Set an 'action taken' on a complaint, so it can be filtered
+    cy.get("#comp-officer-filter").should("exist").click();
+    cy.get(`#${complaintWithActionTakenID}`).should("exist").click();
```
Comment (afwilcox) [inline frontend/cypress/e2e/complaint-search.v2.cy.ts:169]: We have a shortcut function that will allow you navigate directly to a complaint: cy.navigateToDetailsScreen might speed up the test a bit.

File: frontend/src/app/components/containers/complaints/complaint-filter.tsx [hunk 4]
```diff
@@ -320,6 +327,28 @@ export const ComplaintFilter: FC<Props> = ({ type }) => {
             </div>
           </div>
         )}
+        {UserService.hasRole(Roles.CEEB) && (
```
Comment (afwilcox) [inline frontend/src/app/components/containers/complaints/complaint-filter.tsx:330]: Would this be better as a feature flagged component?  

e.g. from call-details.tsx

image

**Review by afwilcox — CHANGES_REQUESTED** [review]

File: backend/src/v1/complaint/complaint.service.ts [hunk 5]
```diff
@@ -1008,13 +1058,30 @@ export class ComplaintService {
...
+        const { data, errors } = await get(token, {
+          query: `{getLeadsByActionTaken (actionCode: "${filters.actionTaken}")}`,
+        });
+        if (errors) {
+          this.logger.error("GraphQL errors:", errors);
+          throw new Error("GraphQL errors occurred");
+        }
+        // If no complaint indentifiers are returned by the CM database, provide a non-empty non-matching list to avoid
+        // the SQL error of IN empty set.
+        const complaintIdentifiers = data.getLeadsByActionTaken.length > 0 ? data.getLeadsByActionTaken : [-1];
```
Comment (afwilcox) [inline backend/src/v1/complaint/complaint.service.ts:1078]: Sorry forgot a comment - looks like we are over our complexity warning for this method.   

Could this be encapsulated into a function and not repeated from earlier (lines: 1006-1031)

File: frontend/src/app/components/containers/complaints/complaint-filter.tsx (hunk 4 above)
Comment (mikevespi) [reply frontend/src/app/components/containers/complaints/complaint-filter.tsx:330]: I don't have a great sense of who needs this filter now vs who will need it in the near future. Does the feature flag approach make more sense here based on what we've heard on that front?

File: backend/src/v1/complaint/complaint.service.ts (hunk 5 above)
Comment (afwilcox) [reply backend/src/v1/complaint/complaint.service.ts]: Arg, we're so close... also it looks like there is one extra item that's been introduced with this change.   See 

https://sonarcloud.io/project/issues?sinceLeakPeriod=true&issueStatuses=OPEN%2CCONFIRMED&pullRequest=708&id=bcgov_nr-compliance-enforcement&open=AZKWOerdTDM-Jdj3eJdr

Not sure if you have another idea for how to simplify that a bit?

**Review by afwilcox — APPROVED** [review]

**Pull request merged**